community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaKey.java

```diff
@@ -0,0 +1,226 @@
+    public long getEntityId()
+    {
+        return entityId;
+    }
+
+    @Override
+    public void setEntityId( long entityId )
+    {
+        this.entityId = entityId;
+    }
+
+    @Override
+    public void from( long entityId, Value... values )
+    {
+        this.entityId = entityId;
+        entityIdIsSpecialTieBreaker = false;
+        assertValidValue( values ).writeTo( this );
+    }
+
+    private TextValue assertValidValue( Value... values )
+    {
+        if ( values.length > 1 )
+        {
+            throw new IllegalArgumentException( "Tried to create composite key with non-composite schema key layout" );
+        }
+        if ( values.length < 1 )
+        {
+            throw new IllegalArgumentException( "Tried to create key without value" );
+        }
+        if ( !Values.isTextValue( values[0] ) )
+        {
+            throw new IllegalArgumentException(
+                    "Key layout does only support strings, tried to create key from " + values[0] );
+        }
+        return (TextValue) values[0];
+    }
+
+    @Override
+    public String propertiesAsString()
+    {
+        return asValue().toString();
+    }
+
+    @Override
+    public Value asValue()
+    {
+        return bytes == null ? Values.NO_VALUE : Values.utf8Value( bytes );
+    }
+
+    // TODO perhaps merge these lowest/highest methods into parent
+    @Override
+    public void initAsLowest()
+    {
+        bytes = null;
+        entityId = Long.MIN_VALUE;
+        entityIdIsSpecialTieBreaker = true;
+    }
+
+    @Override
+    public void initAsHighest()
+    {
+        bytes = null;
+        entityId = Long.MAX_VALUE;
+        entityIdIsSpecialTieBreaker = true;
+    }
+
+    private boolean isHighest()
+    {
+        return entityIdIsSpecialTieBreaker && entityId == Long.MAX_VALUE && bytes == null;
+    }
+
+    /**
+     * Compares the value of this key to that of another key.
+     * This method is expected to be called in scenarios where inconsistent reads may happen (and later retried).
+     *
+     * @param other the {@link StringSchemaKey} to compare to.
+     * @return comparison against the {@code other} {@link StringSchemaKey}.
+     */
+    int compareValueTo( StringSchemaKey other )
+    {
+        // TODO cover all cases of bytes == null and special tie breaker and document
+        if ( bytes != other.bytes )
+        {
+            if ( bytes == null )
+            {
+                return isHighest() ? 1 : -1;
+            }
+            if ( other.bytes == null )
+            {
+                return other.isHighest() ? -1 : 1;
+            }
+        }
+        else
+        {
+            return 0;
+        }
+
+        try
+        {
+            // TODO change to not throw
+            return codePointByteArrayCompare( bytes, other.bytes );
+        }
+        catch ( Exception e )
+        {
+            // We can not throw here because we will visit this method inside a pageCursor.shouldRetry() block.
```

Comment:
not sure if its safe to do a fallback here since it will mean that comparison order is wrong anyway

Reply:
This will be re-written in future work.